File: solutions/typescript/0023.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function mergeKLists(lists: Array<ListNode | null>): ListNode | null {
    const result: number[] = [];
    for (const list of lists) {
        let current = list;
        while (current) {
            result.push(current.val);
            current = current.next;
        }
    }
    result.sort((a, b) => a - b);
    const dummy = new ListNode();
    let current = dummy;
    for (const val of result) {
        current.next = new ListNode(val);
        current = current.next;
    }
    return dummy.next;
};

File: solutions/rust/0023.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn merge_k_lists(lists: Vec<Option<Box<ListNode>>>) -> Option<Box<ListNode>> {
        let mut result = Vec::new();
        for list in lists {
            let mut current = list;
            while let Some(node) = current {
                result.push(node.val);
                current = node.next;
            }
        }
        result.sort();
        let mut dummy = ListNode::new(0);
        let mut current = &mut dummy;
        for val in result {
            current.next = Some(Box::new(ListNode::new(val)));
            current = current.next.as_mut().unwrap();
        }
        dummy.next
    }
}